gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -157,32 +164,112 @@ public void handleResponse(WriteObjectResponse response) {
   }
 
   @Override
-  public void startUpload(InputStream pipeSource) {
-    // Given that the two ends of the pipe must operate asynchronous relative
-    // to each other, we need to start the upload operation on a separate thread.
+  public void startUpload(InputStream pipeSource) throws IOException {
+    return;
+  }
+
+  @Override
+  public boolean isOpen() {
+    return (writeBuffer != null);
+  }
+
+  @Override
+  public synchronized int write(ByteBuffer buffer) throws IOException {
+    checkState(initialized, "initialize() must be invoked before use.");
+    int bytesWritten = 0;
+    if (!isOpen()) {
+      throw new ClosedChannelException();
+    }
+
+    // No point in writing further if upload failed on another thread.
+    if (uploadOperation.isDone()) {
+      waitForCompletionAndThrowIfUploadFailed();
+    }
+
+    while (buffer.remaining() > 0) {
+      int bytesToWrite = 0;
+      // create chunks of MAX_BYTES_PER_MESSAGE and push them into writeBuffer
+      if (lastChunk != null) {
+        bytesToWrite = min(buffer.remaining(), MAX_BYTES_PER_MESSAGE - lastChunk.size());
+        lastChunk = lastChunk.concat(ByteString.copyFrom(buffer, bytesToWrite));
+      } else {
+        bytesToWrite = min(buffer.remaining(), MAX_BYTES_PER_MESSAGE);
+        lastChunk = ByteString.copyFrom(buffer, bytesToWrite);
+      }
+      bytesWritten += bytesToWrite;
+      if (lastChunk.size() == MAX_BYTES_PER_MESSAGE) {
+        try {
+          writeBuffer.put(lastChunk);
+          lastChunk = null;
+        } catch (InterruptedException e) {
+          throw new IOException(
+              String.format(
+                  "Failed to write %d bytes in '%s'", buffer.remaining(), getResourceString()),
+              e);
+        }
+      }
+    }
+
+    return bytesWritten;
+  }
+
+  /** Initialize this channel object for writing. */
+  @Override
+  public void initialize() throws IOException {
+    checkState(!initialized, "initialize() must be invoked only once.");
+    writeBuffer =
+        new LinkedBlockingQueue<ByteString>(
+            Math.toIntExact(channelOptions.getNumberOfBufferedRequests()));
     try {
-      uploadOperation = threadPool.submit(new UploadOperation(pipeSource));
+      uploadOperation = threadPool.submit(new UploadOperation(writeBuffer));
     } catch (Exception e) {
       throw new RuntimeException(String.format("Failed to start upload for '%s'", resourceId), e);
     }
+    initialized = true;
+  }
+
+  @Override
+  public void close() throws IOException {
+    checkState(initialized, "initialize() must be invoked before use.");
+    if (!isOpen()) {
+      return;
+    }
+    if (lastChunk == null) {
+      lastChunk = ByteString.EMPTY;
+    }
+    try {
+      writeBuffer.put(lastChunk);
+      lastChunk = null;
+      handleResponse(waitForCompletionAndThrowIfUploadFailed());
```

Comment:
Should we retry this operation since it would be very expensive to fail here?

Reply:
retry is done by downstream thread. This one just appends to the `blockingQueue` that operation should not fail.